ibis/expr/signature.py

```diff
@@ -0,0 +1,214 @@
+import six
+import itertools
+
+import ibis.util as util
+import ibis.expr.rules as rlz
+
+from ibis.compat import PY2
+from collections import OrderedDict
+
+try:
+    from cytoolz import unique
+except ImportError:
+    from toolz import unique
+
+
+# TODO: could use the primitives defined here to types too
+
+
+_undefined = object()  # marker for missing argument
+
+
+class Argument(object):
+    """Argument definition
+
+    """
+    if PY2:
+        # required to maintain definition order in Annotated metaclass
+        _counter = itertools.count()
+        __slots__ = '_serial', 'validator', 'default'
+    else:
+        __slots__ = 'validator', 'default'
+
+    def __init__(self, validator, default=_undefined):
+        """Argument constructor
+
+        Parameters
+        ----------
+        validator : Union[Callable[[arg], coerced], Type, Tuple[Type]]
+          Function which handles validation and/or coercion of the given
+          argument.
+        default : Union[Any, Callable[[], str]]
+          In case of missing (None) value for validation this will be used.
+          Note, that default value (except for None) must also pass the inner
+          validator.
+          If callable is passed, it will be executed just before the inner, and
+          itsreturn value will be treaded as default.
+        """
+        if PY2:
+            self._serial = next(self._counter)
+
+        self.default = default
+        if isinstance(validator, type):
+            self.validator = rlz.instance_of(validator)
+        elif isinstance(validator, tuple):
+            assert util.all_of(validator, type)
+            self.validator = rlz.instance_of(validator)
+        elif callable(validator):
+            self.validator = validator
+        else:
+            raise TypeError('Argument validator must be a callable, type or '
+                            'tuple of types, given: {}'.format(validator))
+
+    def __eq__(self, other):
+        return (
+            self.validator == other.validator and
+            self.default == other.default
+        )
+
+    def optional(self):
+        return self.default is not _undefined
+
+    def validate(self, value=_undefined, name=None):
+        """
+        Parameters
+        ----------
+        value : Any, default undefined
+          Raises TypeError if argument is mandatory but not value has been
+          given.
+        name : Optional[str]
+          Argument name for error message
+        """
+        if self.optional():
+            if value is _undefined or value is None:
+                if self.default is None:
+                    return None
+                elif util.is_function(self.default):
+                    value = self.default()
```

Comment:
I wasn't aware that we support default argument values that are functions.

Reply:
That is for table `genname`: 

```python
class UnboundTable(PhysicalTable):
    schema = Arg(sch.Schema)
    name = Arg(six.string_types, default=genname)
``` 

I can remove it though.

Reply:
No, let's leave it. It's fine for now.